beetsplug/lastgenre/__init__.py

```diff
@@ -302,42 +325,119 @@ def fetch_track_genre(self, obj):
+
+        Parameters:
+            new_genres (list): The new genre result to process.
+            keep_genres (list): Existing genres to combine with new ones
+            log_label (str): A label (like "track", "album") we possibly
+                combine with a prefix. For example resulting in something like
+                "keep + track" or just "track".
+
+        Returns:
+            tuple: A tuple containing the combined genre string and the
+            'logging label'.
+        """
+        self._log.debug(f"fetched last.fm tags: {new_genres}")
+        combined = deduplicate(keep_genres + new_genres)
+        resolved = self._resolve_genres(combined)
+        reduced = self._to_delimited_genre_string(resolved)
+
+        if new_genres and keep_genres:
+            return reduced, f"keep + {log_label}"
+        if new_genres:
+            return reduced, log_label
+        return None, log_label
+
     def _get_genre(self, obj):
-        """Get the genre string for an Album or Item object based on
-        self.sources. Return a `(genre, source)` pair. The
-        prioritization order is:
+        """Get the final genre string for an Album or Item object
+
+        `self.sources` specifies allowed genre sources, prioritized as follows:
             - track (for Items only)
             - album
             - artist
             - original
             - fallback
             - None
+
+        Parameters:
+            obj: Either an Album or Item object.
+
+         Returns:
+             tuple: A `(genre, label)` pair, where `label` is a string used for
+                logging that describes the result. For example, "keep + artist"
+                indicates that existing genres were combined with new last.fm
+                genres, while "artist" means only new last.fm genres are
+                included.
         """
 
-        # Shortcut to existing genre if not forcing.
-        if not self.config["force"] and self._is_allowed(obj.genre):
-            return obj.genre, "keep"
+        separator = self.config["separator"].get()
+        keep_genres = []
+
+        genres = self._get_existing_genres(obj, separator)
+        if genres and not self.config["force"]:
+            # Without force we don't touch pre-populated tags and return early
+            # with the original contents. We format back to string tough.
+            keep_genres = self._dedup_genres(genres)
+            return separator.join(keep_genres), "keep"
+
+        if self.config["force"]:
+            # Simply forcing doesn't keep any.
+            keep_genres = []
+            # keep_existing remembers, according to the whitelist setting,
+            # any or just allowed genres.
+            if self.config["keep_existing"] and self.config["whitelist"]:
+                keep_genres = self._dedup_genres(genres, whitelist_only=True)
+            elif self.config["keep_existing"]:
+                keep_genres = self._dedup_genres(genres)
```

Comment:
Does `_dedup_genres` need to be called on these genres here? They will be sent to `_combine_genres` -> `_resolve_genres` later on in this function which will achieve the same thing, I think?

Reply:
Yes there were redundant calls of _dedup_genres. All fixed now.